packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -239,14 +240,40 @@ async function redeemSharingLink(
 				const encodedShareUrl = getEncodedShareUrl(
 					odspResolvedUrl.shareLinkInfo?.sharingLinkToRedeem,
 				);
-				const redeemUrl = `${odspResolvedUrl.siteUrl}/_api/v2.0/shares/${encodedShareUrl}`;
-				const { url, headers } = getUrlAndHeadersWithAuth(
-					redeemUrl,
-					storageToken,
-					forceAccessTokenViaAuthorizationHeader,
-				);
-				headers.prefer = "redeemSharingLink";
-				return fetchAndParseAsJSONHelper(url, { headers });
+
+				let redeemUrl: string | undefined;
+				async function callSharesAPI(baseUrl: string): Promise<void> {
+					redeemUrl = `${baseUrl}/_api/v2.0/shares/${encodedShareUrl}`;
+					const { url, headers } = getUrlAndHeadersWithAuth(
+						redeemUrl,
+						storageToken,
+						forceAccessTokenViaAuthorizationHeader,
+					);
+					headers.prefer = "redeemSharingLink";
+					await fetchAndParseAsJSONHelper(url, { headers });
+				}
+
+				const disableUsingTenantDomain = loggerToMonitoringContext(
+					logger,
+				).config.getBoolean("Fluid.Driver.Odsp.DisableUsingTenantDomainForSharesApi");
+				// There is an issue where if we use the siteUrl in /shares, then the allowed length of url is just a few hundred characters
+				// and we fail to do the redeem. But if we use the tenant domain in the url, then the allowed length becomes 2048. So, first
+				// construct the url for /shares using tenant domain but to be on safer side, fallback to what we were using before.
+				if (!disableUsingTenantDomain) {
+					try {
+						await callSharesAPI(new URL(odspResolvedUrl.siteUrl).origin);
```

Comment:
I think you want to move getWithRetryForTokenRefresh() into callSharesAPI(). As code is structured right now, if we have expired token, we will first try two calls, and only then we will refresh token. The flow should be reversed - each call should retry with fresh token.